tracing-forest/src/printer/mod.rs

```diff
@@ -173,3 +173,33 @@ where
         }
     }
 }
+
+/// A [`Processor`] that captures logs during tests and allows them to be presented
+/// when --nocapture is used.
+#[derive(Clone, Debug)]
+pub struct TestCapturePrinter<F> {
+    formatter: F,
+}
+
+impl TestCapturePrinter<Pretty> {
+    /// Construct a new test capturing printer with the default `Pretty` formatter. This printer
+    /// is intented for use in tests only as it works with the default rust stdout capture mechanism
+    pub const fn new() -> Self {
+        TestCapturePrinter { formatter: Pretty }
+    }
+}
+
+impl<F> Processor for TestCapturePrinter<F>
+where
+    F: 'static + Formatter,
+{
+    fn process(&self, tree: Tree) -> processor::Result {
+        let string = match self.formatter.fmt(&tree) {
+            Ok(s) => s,
+            Err(e) => return Err(processor::error(tree, e.into())),
+        };
+
+        print!("{}", string);
```

Comment:
Are you sure not `println!`?

Reply:
Yes, because the pretty printer adds it's own new lines, and stdout/stderr above don't use write ln, they just dump the buffer. As an example with the print!()

```
557ec7fa-3db0-4268-87d5-2877c3bdb280 TRACE       ┝━ 📍 [trace]:  | schema_reload: Cell { value: false } | acp_reload: Cell { value: true } | oauth2_reload: Cell { value: false } | domain_reload: Cell { value: true }
557ec7fa-3db0-4268-87d5-2877c3bdb280 TRACE       ┕━ 📍 [trace]: Create operation success
00000000-0000-0000-0000-000000000000 TRACE    📍 [trace]: internal_migrate_or_create_str -> result | res: Ok(())
938316ff-7c48-4b32-9a45-517e7093c8ec DEBUG    server::internal_migrate_or_create_str [ 2.82ms | 1.89% / 100.00% ]
938316ff-7c48-4b32-9a45-517e7093c8ec DEBUG    ┝━ from_proto_entry_str [ 190µs | 6.73% ]
```